lib/Doctrine/ORM/Utility/PersisterHelper.php

```diff
@@ -0,0 +1,141 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\ORM\Utility;
+
+use Doctrine\Common\Util\ClassUtils;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\ORM\Query\QueryException;
+
+/**
+ * The PersisterHelper contains logic to infer binding types which is used in
+ * several persisters.
+ *
+ * @link   www.doctrine-project.org
+ * @since  2.5
+ * @author Jasper N. Brouwer <jasper@nerdsweide.nl>
+ */
+class PersisterHelper
+{
+    /**
+     * @param mixed                  $value
+     * @param EntityManagerInterface $em
+     *
+     * @return mixed
+     */
+    public static function getValue($value, EntityManagerInterface $em)
+    {
+        if ( ! is_array($value)) {
+            return self::getIndividualValue($value, $em);
+        }
+
+        $newValue = array();
+
+        foreach ($value as $fieldName => $fieldValue) {
+            $newValue[$fieldName] = self::getIndividualValue($fieldValue, $em);
+        }
+
+        return $newValue;
+    }
+
+    /**
+     * @param mixed                  $value
+     * @param EntityManagerInterface $em
+     *
+     * @return mixed
+     */
+    private static function getIndividualValue($value, EntityManagerInterface $em)
+    {
+        if ( ! is_object($value) || ! $em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))) {
+            return $value;
+        }
+
+        return $em->getUnitOfWork()->getSingleIdentifierValue($value);
+    }
+
+    /**
+     * @param string                 $fieldName
+     * @param ClassMetadata          $class
+     * @param EntityManagerInterface $em
+     *
+     * @return string|null
+     *
+     * @throws QueryException
+     */
+    public static function getTypeOfField($fieldName, ClassMetadata $class, EntityManagerInterface $em)
+    {
+        if (isset($class->fieldMappings[$fieldName])) {
+            return $class->fieldMappings[$fieldName]['type'];
+        }
+
+        if ( ! isset($class->associationMappings[$fieldName])) {
+            return null;
+        }
+
+        $assoc = $class->associationMappings[$fieldName];
+
+        if (count($assoc['sourceToTargetKeyColumns']) > 1) {
+            throw QueryException::associationPathCompositeKeyNotSupported();
```

Comment:
I simply moved this from the `BasicEntityPersister`. I'm not sure creating a new Exception with the same meaning would be the way to go. Do you have any suggestions?